sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -454,6 +491,48 @@ case class BroadcastHashJoinExec(
     val (matched, checkCondition, _) = getJoinCondition(ctx, input)
     val numOutput = metricTerm(ctx, "numOutputRows")
 
+    // fast stop if isOriginalInputEmpty = true, should accept all rows in streamedSide
+    if (broadcastRelation.value.isOriginalInputEmpty) {
+      return s"""
+                |// Anti Join isOriginalInputEmpty(true) accept all
+                |$numOutput.add(1);
+                |${consume(ctx, input)}
+          """.stripMargin
+    }
+
+    if (isNullAwareAntiJoin) {
+      if (broadcastRelation.value.allNullColumnKeyExistsInOriginalInput) {
+        return s"""
+                  |// NAAJ
```

Comment:
It would be awesome to verify using org.apache.spark.sql.execution.debug.DebugQuery#debugCodegen that this case does not generate an empty spin-loop -- ie a loop with no executable body.

Reply:
normally it would at least have scan part codegen, let me dump generated-code for you just to make sure.

Reply:
```
/* 026 */   protected void processNext() throws java.io.IOException {
/* 027 */     while ( localtablescan_input_0.hasNext()) {
/* 028 */       InternalRow localtablescan_row_0 = (InternalRow) localtablescan_input_0.next();
/* 029 */       ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 030 */       boolean localtablescan_isNull_0 = localtablescan_row_0.isNullAt(0);
/* 031 */       int localtablescan_value_0 = localtablescan_isNull_0 ?
/* 032 */       -1 : (localtablescan_row_0.getInt(0));
/* 033 */
/* 034 */       // NAAJ
/* 035 */       // isOriginalInputEmpty(false) allNullColumnKeyExistsInOriginalInput(true)
/* 036 */       // reject all
/* 037 */       if (shouldStop()) return;
/* 038 */     }
/* 039 */   }
```

Reply:
what's your suggestion towards to this generated code

Reply:
cc: @cloud-fan ... isn't there a way to "Break" the loop here ? It seems like this would just read and discard the entire input for no reason .. which seems wasteful. Perhaps we should set the stopping condition here ? I don't know enough about gencode to comment further.

Reply:
let me find if there is a way of fast stop the loop.

Reply:
```
from the interface of BufferedRowIterator

normally after a row is appended, it will break the while loop
it will not break until it eventually generate a row, in this case, i don't think we are able to do fast stop

/* 043 */       bhj_mutableStateArray_0[0].zeroOutNullBytes();
/* 044 */
/* 045 */       if (localtablescan_isNull_0) {
/* 046 */         bhj_mutableStateArray_0[0].setNullAt(0);
/* 047 */       } else {
/* 048 */         bhj_mutableStateArray_0[0].write(0, localtablescan_value_0);
/* 049 */       }
/* 050 */
/* 051 */       bhj_mutableStateArray_0[0].write(1, localtablescan_value_1, 2, 1);
/* 052 */       append((bhj_mutableStateArray_0[0].getRow()));
/* 053 */       if (shouldStop()) return;
/* 054 */     }

```

Reply:
```
public boolean hasNext() throws IOException {
    if (currentRows.isEmpty()) {
      processNext();
    }
    return !currentRows.isEmpty();
  }

how about we put a "break"  after
/* 034 */       // NAAJ
/* 035 */       // isOriginalInputEmpty(false) allNullColumnKeyExistsInOriginalInput(true)
/* 036 */       // reject all
                     // break;  // HERE

it this case we break the loop, and the hasNext Check can also works.
```

Reply:
```

see break keyword in basicPhysicalOperators.scala.
it might be ok for NAAJ to use break too. let me do the update and test.

 while ($loopCondition) {
   if ($nextIndex == $batchEnd) {
     long $nextBatchTodo;
     if ($numElementsTodo > ${batchSize}L) {
       $nextBatchTodo = ${batchSize}L;
       $numElementsTodo -= ${batchSize}L;
     } else {
       $nextBatchTodo = $numElementsTodo;
       $numElementsTodo = 0;
       if ($nextBatchTodo == 0) break;
     }
     $batchEnd += $nextBatchTodo * ${step}L;
   }
```

Reply:
```
however if there is a aggr after BHJ

// the break we added here would be wrong. I think this fast stop is not that easy.
@agrawaldevesh how about I hold this issue as "Resolve Later", so i will do more code exploring and resolve this issue in next PR "multi column support",  is that ok for you?

/* 035 */   private void agg_doAggregateWithoutKey_0() throws java.io.IOException {
/* 036 */     // initialize aggregation buffer
/* 037 */     agg_bufIsNull_0 = false;
/* 038 */     agg_bufValue_0 = 0L;
/* 039 */
/* 040 */     while ( localtablescan_input_0.hasNext()) {
/* 041 */       InternalRow localtablescan_row_0 = (InternalRow) localtablescan_input_0.next();
/* 042 */       ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 043 */       boolean localtablescan_isNull_0 = localtablescan_row_0.isNullAt(0);
/* 044 */       int localtablescan_value_0 = localtablescan_isNull_0 ?
/* 045 */       -1 : (localtablescan_row_0.getInt(0));
/* 046 */
/* 047 */       bhj_doConsume_0(localtablescan_value_0, localtablescan_isNull_0);
/* 048 */       // shouldStop check is eliminated
/* 049 */     }
/* 050 */
/* 051 */   }
/* 052 */
/* 053 */   private void bhj_doConsume_0(int bhj_expr_0_0, boolean bhj_exprIsNull_0_0) throws java.io.IOException {
/* 054 */     // NAAJ
/* 055 */     // isOriginalInputEmpty(false) allNullColumnKeyExistsInOriginalInput(true)
/* 056 */     // reject all
/* 057 */     break;  !!!!!!!!!!!
/* 058 */
/* 059 */   }

```

Reply:
It's hard to do early stop with whole-stage-codegen. `CodegenSupport.limitNotReachedChecks` is an example about how to do it for the limit operator, which needs early stop as well.

I'm not sure if it's worth to amend the whole-stage-codegen framework for NAAJ. One choice is to optimize it in AQE: when we find the build side of NAAJ is empty, replace the join node with the stream side plan.

Reply:
Thanks for teaching us about this @cloud-fan !. Given how hard this, let's table this for future. 

@leanken, Thanks for persisting in trying to find a solution for this. I learnt something from your efforts :-). But I agree that this is a "Resolve Later :-P".

Reply:
> Thanks for teaching us about this @cloud-fan !. Given how hard this, let's table this for future.
> 
> @leanken, Thanks for persisting in trying to find a solution for this. I learnt something from your efforts :-). But I agree that this is a "Resolve Later :-P".

@cloud-fan @agrawaldevesh So where are we now? since no further comments for me and test & build is done. Is this RP ready to merge, or is there something i need to follow up, this is my first contribution, i have no idea what's the normal procedure ^_^